Formula/grpcui.rb

```diff
@@ -0,0 +1,22 @@
+class Grpcui < Formula
+  desc "An interactive web UI for gRPC, along the lines of postman"
+  homepage "https://github.com/fullstorydev/grpcui"
+  url "https://github.com/fullstorydev/grpcui/archive/v1.0.0.tar.gz"
+  sha256 "2169f72d23a5e4caceee3b863220bf6a4b4a15f7d88d5bb0de909ca4299a92df"
+
+  depends_on "go" => :build
+
+  def install
+    ENV["GOPATH"] = HOMEBREW_CACHE/"go_cache"
+    (buildpath/"src/github.com/fullstorydev/grpcui").install buildpath.children
+    cd "src/github.com/fullstorydev/grpcui/cmd" do
+      system "go", "build", "-ldflags", "-X main.version=#{version}",
+             "-o", bin/"grpcui", "./..."
+      prefix.install_metafiles
+    end
+  end
+
+  test do
+    system "#{bin}/grpcui", "-help"
```

Comment:
We need a test that exercises the some of the functionality of the app. Version checks or usage checks (foo --version or foo --help) are not sufficient, as explained in the formula cookbook.

In most cases, a good test would involve running a simple test case: run #{bin}/foo input.txt.

- Then you can check that the output is as expected (with assert_equal or assert_match on the output of shell_output)
- You can also check that an output file was created, if that is expected: assert_predicate testpath/"output.txt", :exist?

Some advice for specific cases:

- If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
- If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
- If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
- Same if the software requires a virtual machine, docker instance, etc. to be running.

Reply:
This falls into the case of a UI. But, unfortunately, it does not have a function that runs as command-line only. So I've changed it so the test instead provides bad arguments and asserts that we get the expected error.

Reply:
If it's a web ui you can do a curl request to it.

Reply:
It's a web UI that requires connecting to another server first. So the issue is how to start the UI without another server deterministically available. If the test should actually start the other server, do you have examples? My concerns would be (1) what is the context of the test and how can it download/find an appropriate binary to run as that server? (2) it would require spinning up a separate background process (the server) and would need more complicated code to do process management.

Do you have any examples/suggestions for such a case?

Reply:
@SMillerDev, ping ^^

Reply:
@SMillerDev, did you see my last reply?

Would you expect this test to download and run another program that it can use as the server? Or should it try to connect to a well-known hostname (which could make the test flaky and also make it require internet connectivity)? It cannot operate without a server to connect to and it has no command-line only functions.

Reply:
Try launching `fortio`'s echo server as a local gRPC test server, then running `grpcui` to connect to it, then `curl` to `grpcui` and search the output for an expected fragment (perhaps a trailing `</html>`).

Add `depends_on X => :test` declaration for `fortio` and `curl`, and see the `fortio` test block for hints on how to manage the two server processes.

Reply:
Thanks, @gromgit!

Reply:
@jhump , did you manage to get the test I suggested done?

Reply:
Not yet. I will follow up with that.